### PR TITLE
Upcast log probs in fp32 to avoid `NaN` reward

### PR DESCRIPTION
**Pull request by younesbelkada**
Fixes https://github.com/huggingface/trl/issues/878

As stated by the contributors, it seems the NaN issues disappears when training the model in bf16 / fp32 - hence I believe the fix is to upcast the logprobs in fp32 before computing the rewards to avoid underflow/overflow issues

This should have no effect on training as the reward is not used for backprop

cc @lvwerra

### CODE REVIEW

**Comment by younesbelkada** [comment]
@lewtun what do you think?

**Review by lewtun — COMMENTED** [review]
Seems like a sensible change to me @younesbelkada - thanks for adding it! Before we merge, can we create an integration / unit test that reproduces the issue in indeed just for `fp16` and `bf16` isn't affected?

I realise our CI doesn't currently run on GPU, but even having a slow test we can check manually would be good IMO

**Comment by younesbelkada** [comment]
Yeah makes sense, let me try to do that

**Review by lvwerra — APPROVED** [review]

**Comment by lvwerra** [comment]
also cc @kashif

**Comment by kashif** [comment]
I believe I fixed this issue by removing the cast of the logits to `float()` now the logits are in the actual type of the model: https://github.com/huggingface/trl/pull/885/files#diff-5bbdb5d54108f2162b47bc54dc23c7b8e7744d2941118e60a44c161a4acc0ee8L532-R860

**Comment by lvwerra** [comment]
Then maybe we can close the issue for now until someone reports the issue again?

**Comment by kashif** [comment]
yes lets close it for now
